### DISCUSSION TITLE
Read only mode in IDE

### DISCUSSION THREAD

**Discussion by vitvakatu** [enso-org/enso#6185] [Work In Progress - Engine] [answered]
Hi! We want to discuss the details for the `read-only` mode that would be enabled after clicking on the Workflow-level Play button (see [design doc](https://github.com/enso-org/design/blob/main/epics/basic-libraries/write-action-control/design.md#workflow-level-play-button)).

Questions for the Engine team (@JaroslavTulach @hubertp @4e6 )
1. Is opening visualizations (attach them) safe during execution? Should we disallow opening visualizations for the user?
2. What about visualizations already opened when pressing the Play button? Should we allow them to close? Do we need to close them all forcefully to avoid negative effects on the execution?
3. Do we need to disallow moving nodes in `read-only` mode? Moving nodes will update the project's metadata, which would cause the engine to consider it a code edit. Therefore we should forbid moving nodes. Is it correct?

**Comment by 4e6** [comment, marked as answer]
1,2 Attaching or detaching visualizations does not interrupt the running program. Attaching may cause program re-execution if the visualized value was not cached.
3. All source code changes including changes in the metadata can lead to interruption of the running program and its re-execution. To prevent the interruption and re-execution, the text edit should be sent with `execute: false` flag https://github.com/enso-org/enso/blob/310a2d8ae753f31e6784d512776465706306f56c/docs/language-server/protocol-language-server.md#textapplyedit

**Reply by wdanilo** [reply]
So to answer the @vitvakatu questions from user perspective - users should be able to move nodes, open and close visualizations, but not edit the nodes.

**Reply by farmaazon** [reply]
Wait, but

> Attaching may cause program re-execution if the visualized value was not cached.

So it's not actually _safe_ to attach visualization when we click play button?

**Reply by wdanilo** [reply]
It might cause additional computations in the release mode (play button clicked), that's right. But I don't see why it could be "not safe". What do you mean by that?

**Reply by farmaazon** [reply]
Couldn't the additional computation make nodes with side effects (like making db transactions) run twice?

**Reply by jdunkerley** [reply]
That's what I would be nervous about, I was trying some long-running processes as part of this week's book club and it felt like moving nodes and opening closing visualizations appeared to cause the computation to be aborted and restarted sometimes.

**Reply by vitvakatu** [reply]
So should we disable visualizations attachment, or can the engine handle it? @4e6

**Reply by 4e6** [reply]
I believe, in the scenario described by @jdunkerley the interruption of running computation happened because of the node movements. Node movements produce text edits (editing the metadata) and cause the engine to interrupt and re-execute the program after applying the edits. This can be fixed by adding `execute: false` flag to text edit messages.

Visualization attachment may only cause program execution (without interrupting the currently running one). If we want to prevent the program execution as well, we can add, for example, `execute: false` flag to the `executionContext/attachVisualization` command that will prevent the program execution after attaching the visualization.

**Reply by JaroslavTulach** [reply]
> This can be fixed by adding `execute: false` flag to text edit messages

OK, so the *IDE has full control* on whether an edit is going to lead to re-execution or not. I believe that allows the IDE to send only safe edits that are not going to cause re-execution.

> Visualization attachment may only cause program execution (without interrupting the currently running one)

I believe that additional execution is fine (at least from today's perspective). _Let the IDE send `executionContext/attachVisualization` commands_ for now. Should it cause some problems in the future, we can investigate and decide the best mitigation.

**Reply by farmaazon** [reply]
> Should it cause some problems in the future, we can investigate and decide the best mitigation.

As I wrote, I see a clear problem it may cause: e.g. create doubled database transactions when we run new execution in production mode (or whatever name we currently use for "execution with side effects").

**Reply by hubertp** [reply]
> As I wrote, I see a clear problem it may cause: e.g. create doubled database transactions when we run new execution in production mode (or whatever name we currently use for "execution with side effects").

But changing the execution environment (from `live` to `design` or the other way around) will cancel all pending executions, before re-running things.

**Reply by vitvakatu** [reply]
> But changing the execution environment (from live to design or the other way around) will cancel all pending executions, before re-running things.

We're talking about visualizations attachment, and it can cause double execution, according to @4e6
